src/startup.ts

```diff
@@ -1,60 +1,89 @@
-import { getClient } from "./clients/TorrentClient.js";
-import { MatchMode } from "./constants.js";
 import { CrossSeedError } from "./errors.js";
-import { logger } from "./logger.js";
-import { getRuntimeConfig } from "./runtimeConfig.js";
-import { validateTorrentDir } from "./torrent.js";
+import { Label, logger } from "./logger.js";
+import { Action } from "./constants.js";
 import { validateTorznabUrls } from "./torznab.js";
-import { validateAction } from "./action.js";
+import { getClient } from "./clients/TorrentClient.js";
+import {
+	RuntimeConfig,
+	getRuntimeConfig,
+	setRuntimeConfig,
+} from "./runtimeConfig.js";
+import { existsSync } from "fs";
+import { sep } from "path";
+import { inspect } from "util";
+
+export async function doStartupValidation(
+	zodErrorCount: number
+): Promise<void> {
+	const runtimeConfig: RuntimeConfig = getRuntimeConfig();
+	const validConfigPaths = checkConfigPaths();
+	runtimeConfig.configValid = !zodErrorCount && validConfigPaths;
+	setRuntimeConfig(runtimeConfig);
 
-function validateOptions() {
-	const {
-		action,
-		rtorrentRpcUrl,
-		qbittorrentUrl,
-		transmissionRpcUrl,
-		delugeRpcUrl,
-		dataDirs,
-		linkDir,
-		matchMode,
-		skipRecheck,
-	} = getRuntimeConfig();
-	if (
-		action === "inject" &&
-		!(
-			rtorrentRpcUrl ||
-			qbittorrentUrl ||
-			transmissionRpcUrl ||
-			delugeRpcUrl
-		)
-	) {
+	const downloadClient = getClient();
+	await Promise.all<void>([
+		validateTorznabUrls(),
+		downloadClient?.validateConfig(),
+	]);
+	logger.verbose({
+		label: Label.CONFIGDUMP,
+		message: inspect(runtimeConfig),
+	});
+	if (!runtimeConfig.configValid) {
 		throw new CrossSeedError(
-			"You need to specify --rtorrent-rpc-url, --transmission-rpc-url, --qbittorrent-url, or --deluge-rpc-url when using '-A inject'."
+			`Your configuration is invalid, please see the ${
+				zodErrorCount > 1 || (zodErrorCount > 0 && !validConfigPaths)
+					? "errors"
+					: "error"
+			} above for details.`
 		);
 	}
-	if ((dataDirs && !linkDir) || (!dataDirs && linkDir)) {
-		throw new CrossSeedError(
-			"Data based matching requires both --link-dir and --data-dirs"
-		);
+	logger.info("Your configuration is valid!");
+}
+/**
+ * verifies the config paths provided against the filesystem
+ * @returns true (if paths are valid)
+ */
+function checkConfigPaths(): boolean {
+	const { action, linkDir, dataDirs, torrentDir, outputDir } =
+		getRuntimeConfig();
+
+	let pathFailure = false;
+	if (!existsSync(torrentDir)) {
```

Comment:
are these all directories? maybe we can use `stat().isDirectory()` instead of `exists` to surface errors when people put paths to non-directories

Reply:
Pushed